fedimint-server/src/consensus/mod.rs

```diff
@@ -142,35 +134,43 @@ impl FedimintConsensus {
                 |dbtx| {
                     let consensus_outcome = consensus_outcome.clone();
                     let reference_rejected_txs = reference_rejected_txs.clone();
-                    let peers: BTreeSet<PeerId> = consensus_outcome.contributions.keys().copied().collect();
 
                     Box::pin(async move {
-                        let epoch = consensus_outcome.epoch;
-                        let outcome = consensus_outcome.clone();
-
-                        let UnzipConsensusItem {
-                            epoch_outcome_signature_share: _epoch_outcome_signature_share_cis,
-                            client_config_signature_share: _client_config_signature_share_cis,
-                            transaction: transaction_cis,
-                            consensus_upgrade: consensus_upgrade_cis,
-                            module: module_cis,
-                        } = consensus_outcome
+                        let mut rejected_txs = BTreeSet::new();
+
+                        let items = consensus_outcome.clone()
                             .contributions
                             .into_iter()
-                            .flat_map(|(peer, cis)| cis.into_iter().map(move |ci| (peer, ci)))
-                            .unzip_consensus_item();
-
-                        self.process_module_consensus_items(dbtx, &module_cis, &peers).await;
-                        self.process_upgrade_items(dbtx, &consensus_upgrade_cis).await;
-
-                        let rejected_txs = self
-                            .process_transactions(dbtx, epoch, &transaction_cis)
-                            .await;
+                            .flat_map(|(peer_id, items)| items.into_iter().map(move |item| (peer_id, item)));
+
+                        for (peer_id, consensus_item) in items {
+                            dbtx.set_tx_savepoint()
+                                .await
+                                .expect("Setting transaction savepoint failed");
+
+                            let decision = match self.process_consensus_item(dbtx, consensus_item.clone(), peer_id).await {
+                                Ok(decision) => decision,
+                                Err(error) => {
+                                    warn!(target: "consensus", "Invalid consensus item from {peer_id}: {error}");
+                                    ConsensusDecision::Discard
+                                }
+                            };
+
+                            if decision == ConsensusDecision::Discard {
+                                dbtx.rollback_tx_to_savepoint()
+                                    .await
+                                    .expect("Rolling back transaction to savepoint failed");
```

Comment:
Is this just an added safety or is it expected that modules may have changed the DB even if the item is discarded? Both are ok, just for my understanding.

Reply:
Call: for current modules this is just an additional safety net. But it's easier to always rollback and not just for transactions, so we will make it part of the module interface contract that modules *may* change the DB.